aten/src/ATen/SparseTensorImpl.h

```diff
@@ -191,6 +191,7 @@ struct CAFFE2_API SparseTensorImpl : public TensorImpl {
   // `allow_tensor_metadata_change_` to false would prevent those changes from happening and is
   // undesirable.
   c10::intrusive_ptr<TensorImpl> shallow_copy_and_detach() const override {
+    AT_ASSERT(!is_variable());  // TODO: remove this when Variable and Tensor are merged
```

Comment:
This is a missing check for this function: we don't want users to call `shallow_copy_and_detach()` on a Variable, because Variable stores `sizes_`/`strides_` in its underlying `data_` tensor, and the implementation here will not work for a Variable. Nor do we have a need to call this function on a Variable either, because we are only interested in shallow-copying a non-Variable tensor in `make_variable(...)`.

Reply:
I'm confused by a couple of things:
1) what does a "missing check" mean?
2) How would users call this on a Variable?  This is a method on an Impl -- do you mean Variable::Impl?  I don't even get how you would possibly get here, wouldn't Variable (Impl?) give you the base version, which is not a SparseTensor?

Reply:
Yes by Variable I meant Variable::Impl. This is mostly to prevent people from accidentally doing `sparse_variable.unsafeGetTensorImpl()->shallow_copy_and_detach()` (which shallow-copies the Variable::Impl and is wrong) instead of `sparse_variable.data().unsafeGetTensorImpl()-> shallow_copy_and_detach()` (which shallow-copies the Variable::Impl's underlying `data_` tensor and is correct). We can remove this check if we feel that such misuse won't happen and we are disciplined about what we call `shallow_copy_and_detach()` on.

Reply:
I still don't get how this would be triggered.  If we are trying to prevent `sparse_variable.unsafeGetTensorImpl()->shallow_copy_and_detach()`, but that doesn't call this, how does the assert help?  Shouldn't this be in the base TensorImpl or as an override in Variable::Impl?

Reply:
`sparse_variable.unsafeGetTensorImpl()->shallow_copy_and_detach()` does call this, because this function overrides the one in the base TensorImpl and thus the base TensorImpl one is not called when we call `sparse_variable.unsafeGetTensorImpl()->shallow_copy_and_detach()`. Variable::Impl doesn't override this because shallow-copying a TensorImpl itself is only meaningful for pure TensorImpl, not Variable::Impl.

Reply:
This explanation is incorrect.  I encourage you to try it for yourself and see.

Reply:
Yes sorry it was my mistake - `sparse_variable.unsafeGetTensorImpl()->shallow_copy_and_detach()`  calls the base TensorImpl's `shallow_copy_and_detach()`, and as long as we have the `AT_ASSERT(!is_variable())` check in the base TensorImpl we are in good shape.